lenskit-api/src/main/java/org/lenskit/api/ItemRecommender.java

```diff
@@ -0,0 +1,133 @@
+     *                   universe).
+     * @return The recommended items.
+     * @see #recommend(long, int, Set, Set)
+     */
+    ResultList<Result> recommend(long user, @Nullable Set<Long> candidates);
+
+    /**
+     * Produce a set of recommendations for the user. This is the most general
+     * recommendation method, allowing the recommendations to be constrained by
+     * both a candidate set and an exclude set. The exclude set is applied to
+     * the candidate set, so the final effective candidate set is
+     * <var>canditates</var> minus <var>exclude</var>.
+     *
+     * @param user       The user's ID
+     * @param n          The number of ratings to return. If negative, there is
+     *                   no specific recommendation list size requested.
+     * @param candidates A set of candidate items which can be recommended. If
+     *                   {@code null}, all items are considered candidates.
+     * @param exclude    A set of items to be excluded. If {@code null}, a default
+     *                   exclude set is used.
+     * @return A list of recommended items. If the recommender cannot assign
+     *         meaningful scores, the scores will be {@link Double#NaN}. For
+     *         most scoring recommenders, the items will be ordered in
+     *         decreasing order of score. This is not a hard requirement — e.g.
+     *         set recommenders are allowed to be more flexible.
+     */
+    ResultList<Result> recommend(long user, int n, @Nullable Set<Long> candidates,
+                                 @Nullable Set<Long> exclude);
+
+    /**
+     * Produce a set of recommendations for the user with additional details. This method functions identically to
+     * {@link #recommend(long, int, Set, Set)}, except that it may produce more detailed results. Implementations will
+     * return subclasses of {@link ResultList} that provide access to additional details about each recommendation.
+     *
+     * @param user       The user's ID
+     * @param n          The number of ratings to return. If negative, there is
+     *                   no specific recommendation list size requested.
+     * @param candidates A set of candidate items which can be recommended. If
+     *                   {@code null}, all items are considered candidates.
+     * @param exclude    A set of items to be excluded. If {@code null}, a default
+     *                   exclude set is used.
+     * @return A list of recommended items. If the recommender cannot assign
+     *         meaningful scores, the scores will be {@link Double#NaN}. For
+     *         most scoring recommenders, the items will be ordered in
+     *         decreasing order of score. This is not a hard requirement — e.g.
+     *         set recommenders are allowed to be more flexible.
+     */
+    ResultList<? extends Result> recommendWithDetails(long user, int n, @Nullable Set<Long> candidates,
```

Comment:
I shouldn't say *no* options. There is another option - require individual results to be cast, rather than allowing the scorer/recommender to be cast to gain type-safe automatic access to the results.